File: 2017/js/day6.js
window.AdventOfCode.Day6 = function( input )
{
	input = input.split( /\s/ ).map( ( num ) => parseInt( num, 10 ) );
	
	const duplicates = new Map();
	let part1 = 0;
	let part2 = 0;
	let hit = '';
	
	while( !duplicates.has( hit ) )
	{
		duplicates.set( hit, part1++ );
		
		let highestValue = -1;
		let highestIndex = -1;
		
		for( let i = 0; i < input.length; i++ )
		{
			if( input[ i ] > highestValue )
			{
				highestValue = input[ i ];
				highestIndex = i;
			}
		}
		
		input[ highestIndex ] = 0;
		
		while( highestValue-- > 0 )
		{
			input[ ++highestIndex % input.length ]++;
		}
		
		hit = input.join( ',' );
	}
	
	part2 = part1 - duplicates.get( hit );
	
	return [ part1, part2 ];
};

File: 2022/Answers/Solutions/Day6.cs
using System;

namespace AdventOfCode;

[Answer(6)]
public class Day6 : IAnswer
{
	public (string Part1, string Part2) Solve(string input)
	{
		int SlideThatWindow(int size)
		{
			var start = 0;

			for (var end = 0; end < input.Length; end++)
			{
				var c = input[end];

				for (var j = start; j < end; j++)
				{
					if (c == input[j])
					{
						start = j + 1;
						break;
					}
				}

				if (1 + end - start == size)
				{
					return end + 1;
				}
			}

			return 0;
		}

		var part1 = SlideThatWindow(4);
		var part2 = SlideThatWindow(14);

		return (part1.ToString(), part2.ToString());
	}
}
